torch/nested/_internal/triton_kernels/grouped_mm.py

```diff
@@ -0,0 +1,276 @@
+                "GROUP_SIZE_M": GROUP_SIZE_M,
+            },
+        ]
+    return configs
+
+
+@triton.jit
+def grouped_matmul_kernel(
+    b_ptr,
+    group_size,
+    gn,
+    gk,
+    ldb,
+    a_offsets_ptr,
+    a_ptr,
+    lda,
+    ldc,
+    c_ptr,
+    max_M,
+    # # number of virtual SM
+    # tile sizes
+    BLOCK_SIZE_M: tl.constexpr,
+    BLOCK_SIZE_N: tl.constexpr,
+    BLOCK_SIZE_K: tl.constexpr,
+    GROUP_SIZE_M: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    num_pid_m = tl.cdiv(max_M, BLOCK_SIZE_M)
+    num_pid_n = tl.cdiv(gn, BLOCK_SIZE_N)
+    num_pid_in_group = GROUP_SIZE_M * num_pid_n
+    group_id = pid // num_pid_in_group
+    first_pid_m = group_id * GROUP_SIZE_M
+    group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+    pid_m = first_pid_m + ((pid % num_pid_in_group) % group_size_m)
+    pid_n = (pid % num_pid_in_group) // group_size_m
+    tile_m_idx = pid_m
+    tile_n_idx = pid_n
+
+    # tile_m_idx = tl.program_id(0)
+    # tile_n_idx = tl.program_id(1)
+
+    batch_id = tl.program_id(1)
+
+    # get the gemm size of the current problem
+    a_offset_0 = tl.load(a_offsets_ptr + batch_id, eviction_policy="evict_last").to(
+        tl.int32
+    )
+    a_offset_1 = tl.load(a_offsets_ptr + batch_id + 1, eviction_policy="evict_last").to(
+        tl.int32
+    )
+    gm = a_offset_1 - a_offset_0
+    num_m_tiles = tl.cdiv(gm, BLOCK_SIZE_M)
+    if tile_m_idx < num_m_tiles:
+        # pick up a tile from the current gemm problem
+        # figure out tile coordinates
+
+        offs_am = tile_m_idx * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_bn = tile_n_idx * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+
+        offs_am = offs_am % gm
+        offs_bn = offs_bn % gn
+
+        # Rematerialize on each loop
+        offs_k = tl.arange(0, BLOCK_SIZE_K)
+        a_ptrs = a_ptr + a_offset_0 * lda + (offs_am[:, None]) * lda + offs_k[None, :]
+        b_ptrs = b_ptr + batch_id * gk * gn + offs_k[:, None] * ldb + (offs_bn[None, :])
+        accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
+        for kk in range(0, tl.cdiv(gk, BLOCK_SIZE_K)):
+            # # # hint to Triton compiler to do proper loop pipelining
+            tl.multiple_of(a_ptrs, [16, 16])
+            tl.multiple_of(b_ptrs, [16, 16])
+            a = tl.load(
+                a_ptrs, mask=offs_k[None, :] < gk - kk * BLOCK_SIZE_K, other=0.0
+            )
+            b = tl.load(
+                b_ptrs, mask=offs_k[:, None] < gk - kk * BLOCK_SIZE_K, other=0.0
+            )
+            accumulator += tl.dot(a, b)  # , accumulator) #, "ieee")
+            a_ptrs += BLOCK_SIZE_K
+            b_ptrs += BLOCK_SIZE_K * ldb
+        c = accumulator.to(tl.float16)
+
+        # offs_cm = tile_m_idx * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        # offs_cn = tile_n_idx * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+
+        # Trying to save registers by recomputing indices
```

Comment:
Does it have an impact on perf?

Reply:
It's from https://github.com/pytorch/pytorch/blob/2ab698e708cc1186ca5a4f8e2a722aaba6d9b0da/torch/_inductor/kernel/bmm.py#L110 . It might, but I highly doubt it in comparison to all the other compute going on within this tile and because too many registers can prevent you from scheduling enough work at once.